### PR TITLE
[FIP-38] Tokemak deposit

### PR DESCRIPTION
**Pull request by eswak**
This pull request implements smart contracts for depositing ETH & ERC20s on Tokemak.

- [x] Implement smart contracts
- [x] Unit tests & Mock contracts
- [x] Integration tests on a local mainnet fork
- [x] DAO description
- [x] DAO script

image

### CODE REVIEW

File: contracts/pcv/tokemak/TokemakPCVDepositBase.sol [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    /// @notice event generated when rewards are claimed
+    event ClaimRewards (
+        address indexed _caller,
+        address indexed _token,
+        address indexed _to,
+        uint256 _amount
+    );
+
+    /// @notice the tokemak pool to deposit in
+    address public pool;
```
Comment (Joeysantoro) [inline contracts/pcv/tokemak/TokemakPCVDepositBase.sol:42]: can be immutable

File: contracts/pcv/tokemak/TokemakPCVDepositBase.sol [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+    /// @notice the tokemak rewards contract to claim TOKE incentives
+    address public rewards;
+
+    /// @notice the token stored in the Tokemak pool
+    IERC20 public token;
```
Comment (Joeysantoro) [inline contracts/pcv/tokemak/TokemakPCVDepositBase.sol:48]: can be immutable

File: contracts/pcv/tokemak/TokemakPCVDepositBase.sol [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        address indexed _token,
+        address indexed _to,
+        uint256 _amount
+    );
+
+    /// @notice the tokemak pool to deposit in
+    address public pool;
+
+    /// @notice the tokemak rewards contract to claim TOKE incentives
+    address public rewards;
```
Comment (Joeysantoro) [inline contracts/pcv/tokemak/TokemakPCVDepositBase.sol:45]: probably can be immutable?

File: contracts/pcv/tokemak/TokemakPCVDepositBase.sol [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    /// in anticipation of the execution of a DAO proposal that will call withdraw().
+    /// @dev note that withdraw() calls will revert if this function has not been
+    /// called before.
+    /// @param amountUnderlying of tokens to withdraw in a subsequent withdraw() call.
+    function requestWithdrawal(uint256 amountUnderlying)
+        external
+        onlyGovernorOrAdmin
+        whenNotPaused
+    {
+        ITokemakPool(pool).requestWithdrawal(amountUnderlying);
```
Comment (Joeysantoro) [inline contracts/pcv/tokemak/TokemakPCVDepositBase.sol:86]: Should emit an event

File: contracts/pcv/tokemak/TokemakPCVDepositBase.sol [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    //    a "folder" on IPFS. This message says how much TOKE the account is entitled
+    //    to as their reward (and this is cumulative not just for a single cycle).
+    //    That folder hash is published out to the website which will call out to
+    //    an IPFS gateway, /ipfs/{folderHash}/{account}.json, and get the payload
+    //    they need to submit to the contract. Tx is executed with that payload and
+    //    the account is sent their TOKE."
+    /// For an example of IPFS json file, see :
+    //  https://ipfs.tokemaklabs.xyz/ipfs/Qmf5Vuy7x5t3rMCa6u57hF8AE89KLNkjdxSKjL8USALwYo/0x4eff3562075c5d2d9cb608139ec2fe86907005fa.json
+    function claimRewards(
+        uint256 chainId,
```
Comment (Joeysantoro) [inline contracts/pcv/tokemak/TokemakPCVDepositBase.sol:102]: chainId can probably be hardcoded, is the wallet this address? That can also be hardcoded if yes

File: contracts/pcv/tokemak/EthTokemakPCVDeposit.sol [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+    /// @notice deposit ETH to Tokemak
+    function deposit()
+        external
+        override
+        whenNotPaused
+    {
+        uint256 amount = address(this).balance;
+
+        token.approve(pool, amount);
```
Comment (Joeysantoro) [inline contracts/pcv/tokemak/EthTokemakPCVDeposit.sol:36]: Does this need to be approved?

File: contracts/pcv/tokemak/EthTokemakPCVDeposit.sol [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+        ITokemakEthPool(pool).deposit{value: amount}(amount);
+
+        emit Deposit(msg.sender, amount);
+    }
+
+    /// @notice withdraw tokens from the PCV allocation
+    /// @param amountUnderlying of tokens withdrawn
+    /// @param to the address to send PCV to
+    function withdraw(address to, uint256 amountUnderlying)
```
Comment (Joeysantoro) [inline contracts/pcv/tokemak/EthTokemakPCVDeposit.sol:46]: Is there an edge case around not having requested a withdrawal?

File: test/unit/pcv/ERC20TokemakPCVDeposit.test.ts [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+  beforeEach(async function () {
+    ({ userAddress, pcvControllerAddress, governorAddress } = await getAddresses());
+
+    this.core = await getCore();
+    this.token = await (await ethers.getContractFactory('MockERC20')).deploy();
+    this.toke = await (await ethers.getContractFactory('MockERC20')).deploy();
+    this.pool = await (await ethers.getContractFactory('MockTokemakERC20Pool')).deploy(this.token.address);
+    this.rewards = await (await ethers.getContractFactory('MockTokemakRewards')).deploy(this.toke.address);
+    this.deposit = await (
```
Comment (Joeysantoro) [inline test/unit/pcv/ERC20TokemakPCVDeposit.test.ts:44]: We've been using state variables with local scope to take advantage of typescript types recently on the tests

File: test/unit/pcv/ERC20TokemakPCVDeposit.test.ts [hunk 9]
```diff
@@ -0,0 +1,188 @@
+import { expectRevert, balance, getAddresses, getCore } from '../../helpers';
```
Comment (Joeysantoro) [inline test/unit/pcv/ERC20TokemakPCVDeposit.test.ts:1]: Can use `@test/helpers`. Typescript aliases paths for us

File: test/unit/pcv/ERC20TokemakPCVDeposit.test.ts [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+    this.deposit = await (
+      await ethers.getContractFactory('ERC20TokemakPCVDeposit')
+    ).deploy(this.core.address, this.pool.address, this.rewards.address);
+
+    this.depositAmount = toBN('10000000000000000000'); // 10 token
+  });
+
+  describe('Deposit', function () {
+    it('reverts if paused', async function () {
+      await this.deposit.connect(impersonatedSigners[governorAddress]).pause({});
```
Comment (Joeysantoro) [inline test/unit/pcv/ERC20TokemakPCVDeposit.test.ts:53]: Here and in many places there are empty braces as a param, these can be removed I think

File: test/unit/pcv/EthTokemakPCVDeposit.test.ts [hunk 11]
```diff
@@ -0,0 +1,196 @@
+import { expectRevert, balance, getAddresses, getCore } from '../../helpers';
+import { expect } from 'chai';
+import hre, { ethers } from 'hardhat';
+import { Signer } from 'ethers';
+
+const toBN = ethers.BigNumber.from;
+
+describe('EthTokemakPCVDeposit', function () {
```
Comment (Joeysantoro) [inline test/unit/pcv/EthTokemakPCVDeposit.test.ts:8]: same comments as ERC20 one

File: test/unit/pcv/ERC20TokemakPCVDeposit.test.ts [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+
+      const userBalanceAfter = await this.token.balanceOf(userAddress);
+
+      expect(userBalanceAfter.sub(userBalanceBefore).toString()).to.be.equal(this.depositAmount);
+    });
+  });
+
+  describe('Withdraw ERC20', function () {
+    it('reverts if not PCVController', async function () {
+      await expectRevert(
```
Comment (ditchfieldcaleb) [inline test/unit/pcv/ERC20TokemakPCVDeposit.test.ts:123]: (see other comments about expect revert syntax)

File: test/unit/pcv/ERC20TokemakPCVDeposit.test.ts [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+        'CoreRef: Caller is not a PCV controller'
+      );
+    });
+
+    it('reverts if requestWithdrawal has not been called', async function () {
+      // deposit in the PCVDeposit
+      await this.token.mint(this.deposit.address, this.depositAmount);
+      await this.deposit.deposit();
+
+      await expectRevert(
```
Comment (ditchfieldcaleb) [inline test/unit/pcv/ERC20TokemakPCVDeposit.test.ts:92]: (see other comments about expect revert syntax)

File: test/unit/pcv/ERC20TokemakPCVDeposit.test.ts [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+
+      // Held balance should be 0, now invested into Tokemak
+      expect((await balance.current(this.deposit.address)).toString()).to.be.equal('0');
+    });
+  });
+
+  describe('Withdraw', function () {
+    it('reverts if paused', async function () {
+      await this.deposit.connect(impersonatedSigners[governorAddress]).pause({});
+      await expectRevert(
```
Comment (ditchfieldcaleb) [inline test/unit/pcv/ERC20TokemakPCVDeposit.test.ts:74]: (see other comments about expect revert syntax)

File: test/unit/pcv/ERC20TokemakPCVDeposit.test.ts [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+      await ethers.getContractFactory('ERC20TokemakPCVDeposit')
+    ).deploy(this.core.address, this.pool.address, this.rewards.address);
+
+    this.depositAmount = toBN('10000000000000000000'); // 10 token
+  });
+
+  describe('Deposit', function () {
+    it('reverts if paused', async function () {
+      await this.deposit.connect(impersonatedSigners[governorAddress]).pause({});
+      await expectRevert(this.deposit.deposit(), 'Pausable: paused');
```
Comment (ditchfieldcaleb) [inline test/unit/pcv/ERC20TokemakPCVDeposit.test.ts:54]: (see other comments about expect revert syntax)

File: test/unit/pcv/ERC20TokemakPCVDeposit.test.ts [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+      });
+
+      impersonatedSigners[address] = await ethers.getSigner(address);
+    }
+  });
+
+  beforeEach(async function () {
+    ({ userAddress, pcvControllerAddress, governorAddress } = await getAddresses());
+
+    this.core = await getCore();
```
Comment (ditchfieldcaleb) [inline test/unit/pcv/ERC20TokemakPCVDeposit.test.ts:39]: (see other comment about `this` in test files)

**Comment by Joeysantoro** [comment]
Overall looks really great, you can fix these nits then send to Tokemak team for review. Also feel free to begin DAO scripts

File: contracts/pcv/tokemak/TokemakPCVDepositBase.sol [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+            cycle,
+            wallet,
+            amount
+        );
+
+        ITokemakRewards(rewards).claim(recipient, v, r, s);
+
+        emit ClaimRewards(
+          msg.sender,
+          address(0x2e9d63788249371f1DFC918a52f8d799F4a38C94), // TOKE token
```
Comment (ditchfieldcaleb) [inline contracts/pcv/tokemak/TokemakPCVDepositBase.sol:121]: should declare this as a constant or immutable var instead of hardcoded here

File: contracts/pcv/tokemak/EthTokemakPCVDeposit.sol (hunk 6 above)
Comment (ditchfieldcaleb) [inline contracts/pcv/tokemak/EthTokemakPCVDeposit.sol:36]: What's this approval do/for?

**Review by ditchfieldcaleb — CHANGES_REQUESTED** [review]
Contracts look great overall, just a couple of questions there. Roles all look fine and love the comments.

For the tests I left a lot of comments; most of them are code-style things (we've updated the way we do tests quite a bit in the past few months).

File: contracts/pcv/tokemak/EthTokemakPCVDeposit.sol (hunk 7 above)
Comment (eswak) [reply contracts/pcv/tokemak/EthTokemakPCVDeposit.sol:46]: The Tokemak ETH pool will revert if we have not called `requestWithdrawal` before : https://github.com/Tokemak/tokemak-smart-contracts-public/blob/main/contracts/pools/EthPool.sol#L94-L97

Same for the Tokemak ERC20 pool : https://github.com/Tokemak/tokemak-smart-contracts-public/blob/main/contracts/pools/Pool.sol#L87-L90

File: contracts/pcv/tokemak/EthTokemakPCVDeposit.sol (hunk 6 above)
Comment (eswak) [reply contracts/pcv/tokemak/EthTokemakPCVDeposit.sol:36]: not for the ETH deposit, good catch :)

File: contracts/pcv/tokemak/TokemakPCVDepositBase.sol (hunk 5 above)
Comment (eswak) [reply contracts/pcv/tokemak/TokemakPCVDepositBase.sol:102]: yes, I'll hardcode both

File: test/unit/pcv/ERC20TokemakPCVDeposit.test.ts [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+          '0x4fa17a99b5c319727c9a7c846112902f88d8a261049e70737cd1d60e52609c50', // r
+          '0x44ab278ba36dedbd9b06ceb6c60f884dbaeacb8b3ac4043b901267a2af02ef6f', // s
+          {}
+        ),
+        'Pausable: paused'
+      );
+    });
+
+    it('should claim TOKE rewards & leave them on the deposit', async function () {
+      const rewardAddress = '0x4eff3562075c5d2d9cb608139ec2fe86907005fa';
```
[the remark this answers is not in the document]
Comment (eswak) [reply test/unit/pcv/ERC20TokemakPCVDeposit.test.ts:169]: in the unit test that use the mock rewards contract, it will always work. I removed the address parameter (joey review), so I'll mark this comment as resolved

File: contracts/pcv/tokemak/EthTokemakPCVDeposit.sol (hunk 6 above)
Comment (eswak) [reply contracts/pcv/tokemak/EthTokemakPCVDeposit.sol:36]: removed, WETH does not need approval (pure ETH is passed to the payable function), good catch :)

File: contracts/pcv/tokemak/TokemakPCVDepositBase.sol (hunk 17 above)
Comment (eswak) [reply contracts/pcv/tokemak/TokemakPCVDepositBase.sol:121]: Added as : `address private constant TOKE_TOKEN_ADDRESS = address(0x2e9d63788249371f1DFC918a52f8d799F4a38C94);`

**Review by Joeysantoro — APPROVED** [review]

**Review by ditchfieldcaleb — DISMISSED** [review]

File: proposals/dao/fip_38.ts [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+} from '../../types/types';
+import { TransactionResponse } from '@ethersproject/providers';
+import { expectApprox } from '@test/helpers';
+
+chai.use(CBN(ethers.BigNumber));
+
+// Constants
+const TOKEMAK_REWARDS_ADDRESS = '0x79dD22579112d8a5F7347c5ED7E609e60da713C5';
+const TOKEMAK_WETH_POOL_ADDRESS = '0xD3D13a578a53685B4ac36A1Bab31912D2B2A2F36';
+const TOKEMAK_TOKE_POOL_ADDRESS = '0xa760e26aA76747020171fCF8BdA108dFdE8Eb930';
```
Comment (Joeysantoro) [inline proposals/dao/fip_38.ts:19]: These should probably be added to mainnetAddresses.json right?

File: scripts/deploy/fip_38.ts [hunk 20]
```diff
@@ -0,0 +1,54 @@
+import { ethers } from 'hardhat';
+import { DeployUpgradeFunc, NamedContracts } from '@custom-types/types';
+
+const toBN = ethers.BigNumber.from;
+
+const TOKEMAK_REWARDS_ADDRESS = '0x79dD22579112d8a5F7347c5ED7E609e60da713C5';
```
Comment (Joeysantoro) [inline scripts/deploy/fip_38.ts:6]: This file is not needed

**Comment by Joeysantoro** [comment]
After fixing merge conflicts and pulling in develop should be good to 🚢

File: proposals/dao/fip_38.ts (hunk 19 above)
Comment (eswak) [reply proposals/dao/fip_38.ts:19]: I just needed them here, but I can add them. What `artifactName` do I use ?

**Comment by eswak** [comment]
fixed the e2e tests, didn't know about the `dependencies.ts` file 😅
